Debugger sidebar variables tree expansion states restoration

DenisaCG opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/main/CONTRIBUTING.md
-->

## References
Save the expansion states, set by the user, of the variables in the tree view of the debugger sidebar and restore them along with the layout. 

I'm working to a variant to not change the API.

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes
<!-- Describe the code changes and how they address the issue. -->

The state is stored to be restored when switching from cell-by-cell run and debug step-by-step.
It is also restored when switching between tabs.

## User-facing changes

<!-- Describe any visual or user interaction changes and how they address the issue. -->
Variable tree view expansion is stored
<!-- For visual changes, include before and after screenshots or GIF/mp4/other video demo here. -->

## Backwards-incompatible changes

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->

DenisaCG committed:
store variable tree expansion states

DenisaCG committed:
save variable expansion tree using IStateDB

fcollonval committed:
Make variable tree view uncontrolled

fcollonval committed:
Fix expanding variable in kernel idle mode

krassowski commented:
Thanks for starting this @DenisaCG. I am moving this to 4.2 as we are past feature freeze on 4.1 milestone and this one appears to still be a work in progress. Looking forward to reviewing this one once ready!

fcollonval committed:
Restore expansion state in debug mode

fcollonval committed:
Fix expansion states restoration in "normal" run

fcollonval committed:
Remove debug print

fcollonval committed:
Preserve expansion state in all scenarii

fcollonval committed:
Add test for tree view state preservation

fcollonval committed:
Fix integrity

fcollonval committed:
Refactor sleep

fcollonval committed:
Revert "Refactor sleep"

This reverts commit a178710847a78713a2a38550c785d07a264446e4.

fcollonval committed:
Restore expansion when switching files

fcollonval committed:
Add test for switching tabs

fcollonval committed:
Don't change API - only some behavior

fcollonval committed:
Fix integrity

fcollonval committed:
Don't inspect a variable with existing children

This avoid to loose the view state of children.

It is ok as the children will any way be cleaned if the kernel state changes.

fcollonval committed:
Fix galata helpers

DenisaCG commented:
Thanks for all the work @fcollonval! Everything seems to be working great. 

The expansion tree is stored and restored properly when jumping up and down in the stack or even when closing and opening the `Variables` subpanel or the entire `Debugger` panel itself. 

The only drawback I noticed was that once we refresh the page, the expansion tree is not saved, but everything else from the debugging process is restored as previously left (breakpoints, callstack) . Is this something that would be easy to improve on?

krassowski commented:
It looks like this accumulated a few conflicts, possibly due to https://github.com/jupyterlab/jupyterlab/pull/15375.

krassowski commented:
@DenisaCG and @fco just checking back here as we are nearing 4.2 release - do you have time to resolve the merge conflicts and fix the failing lint check?

krassowski commented:
I moved this to 4.4 as it appears there has not been any recent movement on this, but if you do get to finish it before Monday which is the feature freeze for 4.3 please do let me know and I will change the milestone back!

krassowski commented:
This was not updated in a few years, I will close it for now as it requires extensive rebase to work with the current code. Please feel welcome to open a fresh PR if you would like to add this enhancement.
